src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java

```diff
@@ -0,0 +1,191 @@
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.DetailNode;
+import com.puppycrawl.tools.checkstyle.api.JavadocTokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.JavadocUtil;
+
+/**
+ * <p>
+ * Checks that the
+ * <a href="https://docs.oracle.com/en/java/javase/14/docs/specs/javadoc/doc-comment-spec.html#leading-asterisks">
+ * leading asterisks</a> are aligned vertically under the second asterisk ( &#42; )
+ * of opening Javadoc tag. The alignment of closing Javadoc tag is also checked.
+ * If the line does not contain leading asterisk then that line will be ignored.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code violateExecutionOnNonTightHtml} - Control when to print violations if the
+ * Javadoc being examined by this check violates the tight html rules defined at
+ * <a href="https://checkstyle.org/writingjavadocchecks.html#Tight-HTML_rules">Tight-HTML Rules</a>.
+ * Type is {@code boolean}.
+ * Default value is {@code false}.
+ * </li>
+ * </ul>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code javadoc.asterisk.indentation}
+ * </li>
+ * <li>
+ * {@code javadoc.missed.html.close}
+ * </li>
+ * <li>
+ * {@code javadoc.parse.rule.error}
+ * </li>
+ * <li>
+ * {@code javadoc.unclosedHtml}
+ * </li>
+ * <li>
+ * {@code javadoc.wrong.singleton.html.tag}
+ * </li>
+ * </ul>
+ *
+ * @since 10.15.0
+ */
+@StatelessCheck
+public class JavadocLeadingAsteriskAlignCheck extends AbstractJavadocCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_WRONG_ALIGNMENT = "javadoc.asterisk.indentation";
+
+    @Override
+    public int[] getDefaultJavadocTokens() {
+        return new int[] {
+            JavadocTokenTypes.JAVADOC,
+            JavadocTokenTypes.LEADING_ASTERISK,
+        };
+    }
+
+    @Override
+    public int[] getRequiredJavadocTokens() {
+        return getAcceptableJavadocTokens();
+    }
+
+    @Override
+    public void visitJavadocToken(DetailNode ast) {
```

Comment:
Can we use https://github.com/checkstyle/checkstyle/blob/1bb3d569c975eea323094aed9e3fea4edb5a2e08/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java#L251 to avoid so much tree traversal in the current implementation? We are already "walking" the tree in this check's parent class, I don't think it makes sense to do it all over again in here.

This is what I am thinking:
1. Override beginJavadocTree to handle the JAVADOC token type (we can probably remove this from acceptable tokens?)
2. Store information we need about the alignment of the initial `*`(and possibly for the last line too?)
3. Allow this method to only check the location of a given LEADING_ASTERISK
4. Call finishJavadocTree to reset/clear the alignment information before processing the next javadoc

Reply:
> Can we use
> 
> I don't think it makes sense to do it all over again in here.

Yeah the check implementation is weird and difficult to understand. Checking for the position of comment end token was very difficult as we don't have access to `BLOCK_COMMENT_END` in javadoc checks. That's why the implementation required so much traversing in the tree.


> 1. Override beginJavadocTree

I'm not so familiar with these methods but I'll surely give it a try.

> to handle the JAVADOC token type  (we can probably remove this from acceptable tokens?)

if we remove it from acceptable tokens then how we can get the position of the initial `*` ? I'm assuming that the beginJavadocTree() will get called with JAVADOC token by default? Again I'm not so sure about how beginJavadocTree & finishJavadocTree works so can you guide me please?

> 3. Allow this method to only check the location of a given LEADING_ASTERISK

So there will be only LEADING_ASTERISK token in the acceptable tokens right?

---

I'm okay with your suggestions. But before using beginJavadocTree & finishJavadocTree, can you give me little overview of how javadocBeginTree and finishJavadocTree works? I tried researching about them but it does not seem checkstyle's documentation has covered how they work. @nrmancuso

Reply:
> But before using beginJavadocTree & finishJavadocTree, can you give me little overview of how they works?

I have no idea :) Methods like these are available for all checks but I have not written one that uses them.

Reply:
> Store information we need about the alignment of the initial *(and possibly for the last line too?)

What prevented us from doing this?

Reply:
Do we log violations in begin/finish tree in other checks?

Reply:
yes we do. I couldn't find any checks which logs violation in the begin tree method but there are many checks which logs violation in the finish tree method, here are some of them:

[NoCodeInFileCheck.java](https://github.com/checkstyle/checkstyle/blob/e30260d88bfc5ae5ff110560e28371a8fd8074ba/src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java#L86), [OuterTypeFilenameCheck.java](https://github.com/checkstyle/checkstyle/blob/e30260d88bfc5ae5ff110560e28371a8fd8074ba/src/main/java/com/puppycrawl/tools/checkstyle/checks/OuterTypeFilenameCheck.java#L123), [EqualsHashCodeCheck.java](https://github.com/checkstyle/checkstyle/blob/e30260d88bfc5ae5ff110560e28371a8fd8074ba/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/EqualsHashCodeCheck.java#L166), etc